chain/client/tests/ganache.rs

```diff
@@ -0,0 +1,174 @@
+        *genesis_block.hash(),
+    );
+    env.clients[0].process_tx(tx, false, false);
+    let mut last_block = genesis_block;
+    for i in 1..3 {
+        last_block = env.clients[0].produce_block(i).unwrap().unwrap();
+        env.process_block(0, last_block.clone(), Provenance::PRODUCED);
+    }
+    let query_state = |chain: &mut near_chain::Chain,
+                       runtime_adapter: Arc<dyn RuntimeAdapter>,
+                       account_id: AccountId| {
+        let final_head = chain.store().final_head().unwrap();
+        let last_final_block = chain.get_block(&final_head.last_block_hash).unwrap().clone();
+        println!("{}", last_final_block.header().height());
+        let response = runtime_adapter
+            .query(
+                0,
+                &last_final_block.chunks()[0].prev_state_root(),
+                last_final_block.header().height(),
+                last_final_block.header().raw_timestamp(),
+                &final_head.prev_block_hash,
+                last_final_block.hash(),
+                last_final_block.header().epoch_id(),
+                &QueryRequest::ViewState { account_id, prefix: vec![].into() },
+            )
+            .unwrap();
+        match response.kind {
+            QueryResponseKind::ViewState(view_state_result) => view_state_result.values,
+            // QueryResponseKind::ViewCode(code) => code.code,
+            _ => panic!("Wrong return value"),
+        }
+    };
+
+    let function_call_tx = SignedTransaction::from_actions(
+        2,
+        "test0".to_string(),
+        "test0".to_string(),
+        &signer,
+        vec![Action::FunctionCall(FunctionCallAction {
+            method_name: "write_block_height".to_string(),
+            args: vec![],
+            gas: 100000000000000,
+            deposit: 0,
+        })],
+        *last_block.hash(),
+    );
+    env.clients[0].process_tx(function_call_tx, false, false);
+    for i in 3..9 {
+        last_block = env.clients[0].produce_block(i).unwrap().unwrap();
+        env.process_block(0, last_block.clone(), Provenance::PRODUCED);
+    }
+
+    let runtime_adapter = env.clients[0].runtime_adapter.clone();
+    let state =
+        query_state(&mut env.clients[0].chain, runtime_adapter.clone(), "test0".to_string());
+
+    println!("{:?}", from_base64(&state[0].key).unwrap());
+    env.clients[0]
+        .chain
+        .patch_state(&[StateRecord::Data {
+            account_id: "test0".to_string(),
+            data_key: from_base64(&state[0].key).unwrap(),
+            value: vec![40u8],
+        }])
+        .unwrap();
+
+    for i in 9..20 {
+        last_block = env.clients[0].produce_block(i).unwrap().unwrap();
+        env.process_block(0, last_block.clone(), Provenance::PRODUCED);
+    }
+
+    let runtime_adapter = env.clients[0].runtime_adapter.clone();
+    let state2 =
+        query_state(&mut env.clients[0].chain, runtime_adapter.clone(), "test0".to_string());
+    println!("{:?}", state2);
```

Comment:
Hm, I am having second thoughts about testing here. The current test has this problem. The "essence of test" are this lines:

```rust
        .patch_state(&[StateRecord::Data {
            account_id: "test0".to_string(),
            data_key: from_base64(&state[0].key).unwrap(),
            value: vec![40u8],
        }])

    let state2 =
        query_state(&mut env.clients[0].chain, runtime_adapter.clone(), "test0".to_string());
```

All the rest is setup boilerplate. If all the test would have this essense/boilerplate ratio, then adding new tests would be painfunl. This is OK for draft, but, before merging this, we need to make sure that:

* there are at least two tests
* common setup code is extracted into a function
* the tests themselves are relatively small. 

Rust-analyzer's integration test are a good model here:

https://github.com/rust-analyzer/rust-analyzer/blob/9a431c26f4528e2649de0ca171a38c93e473c94e/crates/rust-analyzer/tests/rust-analyzer/main.rs#L40-L73

Each test uses a DSL to setup context/submit quireis. The interpreter of the DSL is in a separate file. It's [rather gnarly](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/rust-analyzer/tests/rust-analyzer/support.rs#L21-L277), but it isn't repeated for every test.

Reply:
The DSL like approach in rust-analyzer test looks near! I guess this DSL shall be more general to most of `chain/client/tests/*.rs`. patch_state specifically only need one test: init with some state => patch state => check state patch is effective. And it borrows the boilerplate from `chain/client/tests/process_blocks.rs` where many tests there are heavily boilerplated in a similar manner

Reply:
Definitely agree that the problem is pre-existing, and that refactoring of existing tests should be done separately. 

However, I think we can de-boilerplate new tests without necessary touching the old ones. Doing something simple like just extracting the common `setup` function can be done independently. Extending that to support the rest of the tests could be done separately. 

I think that, long term, that's the less risky approach. If you have a bunch of tests you want to refactor to a DSL, there are two ways you can proceed:

* design the DSL upfront and re-write exciting tests to it in one go
* design DSL for *new* test, and iterate on it for a while; move the rest of tests later. 

Second one is usually better, as it allows for a small change and experience-based refinement of the design. 


The more general point here is that I am not a fan of merging functionality which isn't very easy to test. Psychologically, adding tests is not as fun as adding code, so I think it's important to make writing new tests easy, otherwise follow up code fixes might go untested. 

So, I don't think we need to go all in on designing a full test-framework here, but I do think that adding one more test for *this* specific functionality could be made easy by just exctracting some common code in this file.

Reply:
Sounds good! I'll try DSL for the new tests

Reply:
Done